### PR TITLE
remove device sync (torch.where) and vectorize batch dims

### PR DESCRIPTION
**Pull request by nikitaved**
As per title. It is maybe worth adding that assert.

But for now, @tfogal , could you please test it out locally as I am having difficulties with setting up the environment?

### CODE REVIEW

**Review by tfogal — CHANGES_REQUESTED** [review]
Thanks for this!

With thunder, it looks like we live another day with this error---it gets around our issues with torch.where 🎉 !
But it does seem to actually change semantics. In eager mode, the `scatter_` later now fails:

```
[rank0]:     words_embeddings = self.word_embeddings(input_ids)
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1716, in _wrapped_call_impl
[rank0]:     return self._call_impl(*args, **kwargs)
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1727, in _call_impl
[rank0]:     return forward_call(*args, **kwargs)
[rank0]:   File "/home/tfogal/dev/nemo/nemo/collections/multimodal/models/multimodal_llm/neva/neva_model.py", line 155, in forward
[rank0]:     return self.replace_media_embeddings(input_ids, words_embeddings, media)
[rank0]:   File "/home/tfogal/dev/nemo/nemo/collections/multimodal/models/multimodal_llm/neva/neva_model.py", line 240, in replace_media_embeddings
[rank0]:     updated_input_embeds.scatter_(1, padded_media_indices, media_features)
[rank0]: RuntimeError: Expected index [2, 98304, 5120] to be smaller than self [2, 640, 5120] apart from dimension 1 and to be smaller size than src [2, 256, 5120]
```

Could you take a look?

**Comment by tfogal** [comment]
Applying this 'just print metadata' patch:

```diff
diff --git a/nemo/collections/multimodal/models/multimodal_llm/neva/neva_model.py b/nemo/collections/multimoda
l/models/multimodal_llm/neva/neva_model.py
index 1741880bb..adece509e 100644
--- a/nemo/collections/multimodal/models/multimodal_llm/neva/neva_model.py
+++ b/nemo/collections/multimodal/models/multimodal_llm/neva/neva_model.py
@@ -188,50 +188,78 @@ class NevaWordEmbeddingMixin(torch.nn.Module, adapter_mixins.AdapterModuleMixin)
     def replace_media_embeddings(self, input_ids, inputs_embeds, media):
         if media is None:
             return inputs_embeds
+        print("call of replace_media_embeddings(",
+              input_ids.dtype, ":", input_ids.shape, ", ",
+              inputs_embeds.dtype, ":", inputs_embeds.shape, ", ",
+              media.dtype, ":", media.shape, ")")

         batch_size, sequence_length, hidden_size = inputs_embeds.shape

         # calculate media features without gradients
         media_features = self.encode_vision_x(media)  # b T F S(eq) H(idden)
+        print(f"media features dtype:shape = {media_features.dtype}:{media_features.shape}")
         num_images_per_sample = media_features.size(1)
         num_patches = media_features.size(3) * media_features.size(2)
         # flatten patches
         media_features = media_features.view(batch_size, -1, hidden_size)
+        print(f"media features new shape: {media_features.shape}")

         # create an indices matrix used in torch.scatter
         padded_media_indices = torch.ones(
             (batch_size, num_images_per_sample), dtype=torch.long, device=input_ids.device
         )
+        print(f"padded_media_indices dt:shape:", padded_media_indices.dtype,
+              padded_media_indices.shape)
         padded_media_indices *= sequence_length
         for idx, input_id in enumerate(input_ids):
+            print(f"idx {idx} input_id dt:shape = ", input_id.dtype, ":",
+                  input_id.shape)
             media_end_positions = torch.where(input_id == self.media_end_id)[0]
+            print("media end pos=", media_end_positions.dtype, ":",
+                  media_end_positions.shape)
             if self.use_im_start_end:
                 # locate the first media token positions
                 padded_media_indices[idx, : len(media_end_positions)] = media_end_positions - num_patches
+                print("im start end path, PMI=", padded_media_indices.dtype,
+                      ":", padded_media_indices.shape)
                 assert (
                     input_id[padded_media_indices[idx, : len(media_end_positions)] - 1] == self.media_start_id
                 ).all()
             else:
                 padded_media_indices[idx, : len(media_end_positions)] = media_end_positions - num_patches + 1
+                print("no use start/end, PMI=", padded_media_indices.dtype,
+                      ":", padded_media_indices.shape)
                 assert (input_id[padded_media_indices[idx, : len(media_end_positions)]] == self.media_start_id).all()

         # use indices to create a span
         padded_media_indices = padded_media_indices.unsqueeze(-1) + torch.arange(
             num_patches, device=padded_media_indices.device
         ).repeat(*padded_media_indices.shape, 1)
+        print("recompute PMI=", padded_media_indices.dtype,
+              ":", padded_media_indices.shape)
         padded_media_indices = padded_media_indices.reshape(batch_size, -1)
+        print("reshaped PMI=", padded_media_indices.dtype,
+              ":", padded_media_indices.shape)
         padded_media_indices = repeat(padded_media_indices, 'b s -> b s h', h=hidden_size)
+        print("repeated PMI=", padded_media_indices.dtype,
+              ":", padded_media_indices.shape)

         # concat placeholder
         updated_input_embeds = torch.cat(
             (inputs_embeds, torch.zeros((batch_size, num_patches, hidden_size), device=inputs_embeds.device)), dim=1
         )
         updated_input_embeds = updated_input_embeds.type(media_features.dtype)
+        print("updated input embeds=", updated_input_embeds.dtype,
+              ":", updated_input_embeds.shape)
         # scatter media_features
         updated_input_embeds.scatter_(1, padded_media_indices, media_features)
+        print("UIE post scatter_=", updated_input_embeds.dtype,
+              ":", updated_input_embeds.shape)

         # chop off placeholder
         updated_input_embeds = updated_input_embeds[:, :sequence_length]
+        print("UIE after chop=", updated_input_embeds.dtype,
+              ":", updated_input_embeds.shape)

         return updated_input_embeds
```

and running with the `tfogal/thunder-nemo` branch but thunder disabled spams the terminal with a ton of such calls. AFAICT each call produces the same output:

```
call of replace_media_embeddings( torch.int64 : torch.Size([2, 384]) ,  torch.float32 : torch.Size([2, 384, 5120]) ,  torch.float32 : torch.Size([2, 1, 1, 3, 224, 224]) )
media features dtype:shape = torch.float16:torch.Size([2, 1, 1, 256, 5120])
media features new shape: torch.Size([2, 256, 5120])
padded_media_indices dt:shape: torch.int64 torch.Size([2, 1])
idx 0 input_id dt:shape =  torch.int64 : torch.Size([384])
media end pos= torch.int64 : torch.Size([1])
no use start/end, PMI= torch.int64 : torch.Size([2, 1])
idx 1 input_id dt:shape =  torch.int64 : torch.Size([384])
media end pos= torch.int64 : torch.Size([1])
no use start/end, PMI= torch.int64 : torch.Size([2, 1])
recompute PMI= torch.int64 : torch.Size([2, 1, 256])
reshaped PMI= torch.int64 : torch.Size([2, 256])
repeated PMI= torch.int64 : torch.Size([2, 256, 5120])
updated input embeds= torch.float16 : torch.Size([2, 640, 5120])
UIE post scatter_= torch.float16 : torch.Size([2, 640, 5120])
UIE after chop= torch.float16 : torch.Size([2, 384, 5120])
```

**Comment by nikitaved** [comment]
@tfogal , thank you for the info. The shape relation is now more or less clear to me. I have submitted a fix. Let's see how it goes.

File: nemo/collections/multimodal/models/multimodal_llm/neva/neva_model.py [hunk 1]
```diff
@@ -198,22 +198,30 @@ def replace_media_embeddings(self, input_ids, inputs_embeds, media):
...
         media_features = media_features.view(batch_size, -1, hidden_size)
 
-        # create an indices matrix used in torch.scatter
-        padded_media_indices = torch.ones(
-            (batch_size, num_images_per_sample), dtype=torch.long, device=input_ids.device
+        # create an indices matrix used in torch.scatter {
+        sorted_media_end_positions_mask, media_end_positions_mask_sort_idx = (
+            # NOTE: to(torch.long) is needed because PyTorch does not have sort for boolean tensors on CUDA
+            (input_ids == self.media_end_id).to(torch.long).sort(dim=-1, descending=True, stable=True)
         )
```
Comment (nikitaved) [inline nemo/collections/multimodal/models/multimodal_llm/neva/neva_model.py:205]: I have a suspicion that stability might not necessarily be a requirement judging from the code below. Let's test the current impl and then I will try to replace `sort` with `topk(...,k=num_images_per_sample)`.

**Pull request closed without merging**